I can reproduce the failure on Java 8.
`[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.commons.lang3.CharSequenceUtilsTest
[ERROR] Tests run: 14, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.202 s <<< FAILURE! -- in org.apache.commons.lang3.CharSequenceUtilsTest
[ERROR] org.apache.commons.lang3.CharSequenceUtilsTest.testRegionMatches -- Time elapsed: 0.009 s <<< FAILURE!
org.opentest4j.AssertionFailedError: String Failed test 𐐀[0] caseblind 𐐨[0] 2 => true ==> expected: <true> but was: <false>`

The new Deseret test cases are added to the existing regionMatches parity test data, which also checks the String.regionMatches(true, ...) path. On Java 8, String.regionMatches(true, 0, "𐐨", 0, 2) for "𐐀" returns false, so the String check fails before this can purely validate the CharSequenceUtils path.

That seems to make the intended behavior a design question:
@garydgregory 

Should CharSequenceUtils.regionMatches continue to match String.regionMatches behavior on the running JDK, including Java 8, or should it intentionally provide supplementary-code-point case folding even where Java 8 String.regionMatches does not?

If the goal is JDK parity, the new cases may need to be version-gated or adjusted. If the goal is improved Unicode behavior beyond Java 8 String.regionMatches, then these cases may need to move out of the existing String parity test data and into a dedicated CharSequenceUtils/non-String test.

--- a/src/main/java/org/apache/commons/lang3/CharSequenceUtils.java
+++ b/src/main/java/org/apache/commons/lang3/CharSequenceUtils.java
@@ -319,9 +319,14 @@ public class CharSequenceUtils {
             if (!ignoreCase) {
                 return false;
             }
-            // The same case-insensitive check as String#regionMatches(boolean, int, String, int, int),
-            // which also folds case of a supplementary code point split across a surrogate pair.
+            // The same case-insensitive check as String#regionMatches(boolean, int, String, int, int).
             if (!equalsIgnoreCase(c1, c2)) {
+                // String only folds a supplementary code point split across a surrogate pair from Java 9
+                // on; on Java 8 it stays char-by-char. Track the running JDK so every CharSequence type
+                // gives the same result that String does there.
+                if (SystemUtils.IS_JAVA_1_8) {
+                    return false;
+                }
                 int cp1 = c1;
                 if (Character.isHighSurrogate(c1)) {
                     if (index1 + 1 < end1 && Character.isLowSurrogate(cs.charAt(index1 + 1))) {

--- a/src/main/java/org/apache/commons/lang3/CharSequenceUtils.java
+++ b/src/main/java/org/apache/commons/lang3/CharSequenceUtils.java
@@ -27,6 +27,17 @@ public class CharSequenceUtils {
 
     private static final int NOT_FOUND = -1;
 
+    /**
+     * Whether the running JDK folds a supplementary code point split across a surrogate pair when comparing case
+     * insensitively in {@link String#regionMatches(boolean, int, String, int, int)}. JDKs up to and including Java 11
+     * compare surrogate by surrogate and never match such a pair; later JDKs fold the whole code point. Probing what
+     * {@link String} actually does (rather than gating on a version constant) keeps every {@link CharSequence} type in
+     * step with {@link String} on whatever JDK is running. DESERET CAPITAL LETTER LONG I (U+10400) folds to its small
+     * form (U+10428).
+     */
+    private static final boolean STRING_FOLDS_SUPPLEMENTARY_CASE =
+            new String(Character.toChars(0x10400)).regionMatches(true, 0, new String(Character.toChars(0x10428)), 0, 2);
+
     static final int TO_STRING_LIMIT = 16;
 
     private static boolean checkLaterThan1(final CharSequence cs, final CharSequence searchChar, final int len2, final int start1) {
@@ -321,10 +332,9 @@ public class CharSequenceUtils {
             }
             // The same case-insensitive check as String#regionMatches(boolean, int, String, int, int).
             if (!equalsIgnoreCase(c1, c2)) {
-                // String only folds a supplementary code point split across a surrogate pair from Java 9
-                // on; on Java 8 it stays char-by-char. Track the running JDK so every CharSequence type
-                // gives the same result that String does there.
-                if (SystemUtils.IS_JAVA_1_8) {
+                // Only fold a supplementary code point split across a surrogate pair where String itself does, so
+                // every CharSequence type gives the same result that String does on the running JDK (see field).
+                if (!STRING_FOLDS_SUPPLEMENTARY_CASE) {
                     return false;
                 }
                 int cp1 = c1;

--- a/src/test/java/org/apache/commons/lang3/CharSequenceUtilsTest.java
+++ b/src/test/java/org/apache/commons/lang3/CharSequenceUtilsTest.java
@@ -161,13 +161,6 @@ class CharSequenceUtilsTest extends AbstractLangTest {
             new TestData("Abc",  false,      1,     "abc",  1,     2,    true),
             new TestData("Abcd", true,       1,     "abcD", 1,     2,    true),
             new TestData("Abcd", false,      1,     "abcD", 1,     2,    true),
-            // Deseret CAPITAL LONG I (U+10400) folds to SMALL LONG I (U+10428): a supplementary
-            // code point split across a surrogate pair must fold like java.lang.String does.
-            new TestData("\uD801\uDC00",  true,  0, "\uD801\uDC28", 0, 2, true),
-            new TestData("\uD801\uDC00",  false, 0, "\uD801\uDC28", 0, 2, false),
-            new TestData("\uD801\uDC28",  true,  0, "\uD801\uDC00", 0, 2, true),
-            new TestData("x\uD801\uDC00", true,  1, "\uD801\uDC28", 0, 2, true),
-            new TestData("\uD801\uDC00",  true,  0, "\uD801\uDC29", 0, 2, false),
             // @formatter:on
     };
 
@@ -281,6 +274,22 @@ class CharSequenceUtilsTest extends AbstractLangTest {
                 }
             }.run(data, "CSNonString");
         }
+    }
+
+    /**
+     * The green path for a non-String CharSequence must fold a supplementary code point as one unit, not as two bare
+     * surrogates, so the case-insensitive result no longer depends on the argument's runtime type. Deseret CAPITAL LONG I
+     * (U+10400) folds to SMALL LONG I (U+10428). These cases are kept out of {@link #TEST_DATA} because
+     * {@link String#regionMatches(boolean, int, String, int, int)} only folds supplementary code points from Java 9
+     * onward, so the shared String-parity rows would fail on Java 8.
+     */
+    @Test
+    void testRegionMatchesSupplementaryCaseFold() {
+        assertTrue(CharSequenceUtils.regionMatches(new StringBuilder("\uD801\uDC00"), true, 0, "\uD801\uDC28", 0, 2));
+        assertFalse(CharSequenceUtils.regionMatches(new StringBuilder("\uD801\uDC00"), false, 0, "\uD801\uDC28", 0, 2));
+        assertTrue(CharSequenceUtils.regionMatches(new StringBuilder("\uD801\uDC28"), true, 0, "\uD801\uDC00", 0, 2));
+        assertTrue(CharSequenceUtils.regionMatches(new StringBuilder("x\uD801\uDC00"), true, 1, "\uD801\uDC28", 0, 2));
+        assertFalse(CharSequenceUtils.regionMatches(new StringBuilder("\uD801\uDC00"), true, 0, "\uD801\uDC29", 0, 2));
     }
 
     @Test

--- a/src/test/java/org/apache/commons/lang3/CharSequenceUtilsTest.java
+++ b/src/test/java/org/apache/commons/lang3/CharSequenceUtilsTest.java
@@ -28,6 +28,7 @@ import static org.junit.jupiter.params.provider.Arguments.arguments;
 
 import java.lang.reflect.Constructor;
 import java.lang.reflect.Modifier;
+import java.nio.CharBuffer;
 import java.util.Random;
 import java.util.stream.IntStream;
 import java.util.stream.Stream;
@@ -276,20 +277,31 @@ class CharSequenceUtilsTest extends AbstractLangTest {
         }
     }
 
+    private static void assertRegionMatchesParity(final String source, final boolean ignoreCase, final int toffset, final String other,
+            final int ooffset, final int len) {
+        // String is the reference: whatever the running JDK does for String, every CharSequence type must match.
+        final boolean expected = source.regionMatches(ignoreCase, toffset, other, ooffset, len);
+        final CharSequence[] sources = {source, new StringBuilder(source), new StringBuffer(source), CharBuffer.wrap(source)};
+        for (final CharSequence cs : sources) {
+            assertEquals(expected, CharSequenceUtils.regionMatches(cs, ignoreCase, toffset, other, ooffset, len),
+                    cs.getClass().getSimpleName() + " differs from String for " + source + " vs " + other);
+        }
+    }
+
     /**
-     * The green path for a non-String CharSequence must fold a supplementary code point as one unit, not as two bare
-     * surrogates, so the case-insensitive result no longer depends on the argument's runtime type. Deseret CAPITAL LONG I
-     * (U+10400) folds to SMALL LONG I (U+10428). These cases are kept out of {@link #TEST_DATA} because
-     * {@link String#regionMatches(boolean, int, String, int, int)} only folds supplementary code points from Java 9
-     * onward, so the shared String-parity rows would fail on Java 8.
+     * A supplementary code point split across a surrogate pair must fold the same way for every {@link CharSequence}
+     * type that it does for {@link String} on the running JDK. {@link String#regionMatches(boolean, int, String, int, int)}
+     * only folds such a code point from Java 9 on, so these rows are checked against {@link String} itself rather than a
+     * fixed result: {@link String}, {@link StringBuilder}, {@link StringBuffer} and {@link CharBuffer} all have to agree.
+     * Deseret CAPITAL LONG I (U+10400) folds to SMALL LONG I (U+10428).
      */
     @Test
     void testRegionMatchesSupplementaryCaseFold() {
-        assertTrue(CharSequenceUtils.regionMatches(new StringBuilder("\uD801\uDC00"), true, 0, "\uD801\uDC28", 0, 2));
-        assertFalse(CharSequenceUtils.regionMatches(new StringBuilder("\uD801\uDC00"), false, 0, "\uD801\uDC28", 0, 2));
-        assertTrue(CharSequenceUtils.regionMatches(new StringBuilder("\uD801\uDC28"), true, 0, "\uD801\uDC00", 0, 2));
-        assertTrue(CharSequenceUtils.regionMatches(new StringBuilder("x\uD801\uDC00"), true, 1, "\uD801\uDC28", 0, 2));
-        assertFalse(CharSequenceUtils.regionMatches(new StringBuilder("\uD801\uDC00"), true, 0, "\uD801\uDC29", 0, 2));
+        assertRegionMatchesParity("\uD801\uDC00", true, 0, "\uD801\uDC28", 0, 2);
+        assertRegionMatchesParity("\uD801\uDC00", false, 0, "\uD801\uDC28", 0, 2);
+        assertRegionMatchesParity("\uD801\uDC28", true, 0, "\uD801\uDC00", 0, 2);
+        assertRegionMatchesParity("x\uD801\uDC00", true, 1, "\uD801\uDC28", 0, 2);
+        assertRegionMatchesParity("\uD801\uDC00", true, 0, "\uD801\uDC29", 0, 2);
     }
 
     @Test

